src/module.js

```diff
@@ -79,24 +79,30 @@ function processModule( name, options, executeNow, modifiers = {} ) {
 		after: setHookFunction( module, "after" )
 	};
 
-	const currentModule = config.currentModule;
+	const prevModule = config.currentModule;
+	config.currentModule = module;
+
 	if ( objectType( executeNow ) === "function" ) {
 		moduleStack.push( module );
-		config.currentModule = module;
 
-		const cbReturnValue = executeNow.call( module.testEnvironment, moduleFns );
-		if ( cbReturnValue != null && objectType( cbReturnValue.then ) === "function" ) {
-			Logger.warn( "Returning a promise from a module callback is not supported. " +
-				"Instead, use hooks for async behavior. " +
-				"This will become an error in QUnit 3.0." );
+		try {
+			const cbReturnValue = executeNow.call( module.testEnvironment, moduleFns );
+			if ( cbReturnValue != null && objectType( cbReturnValue.then ) === "function" ) {
+				Logger.warn( "Returning a promise from a module callback is not supported. " +
+					"Instead, use hooks for async behavior. " +
+					"This will become an error in QUnit 3.0." );
+			}
+		} finally {
+
+			// If the module closure threw an uncaught error during the load phase,
+			// we let this bubble up to global error handlers. But, not until after
+			// we teardown internal state to ensure correct module nesting.
+			// Ref https://github.com/qunitjs/qunit/issues/1478.
+			moduleStack.pop();
+			config.currentModule = module.parentModule || prevModule;
```

Comment:
Same as before, we only restore `currentModule` if the module in question is a scoped module. If it is a linear module, then we assign current and leave it as-is. I hope the new structure makes this more obvious. In the old code, this was done a bit strangely.